daml-script/runner/src/main/scala/com/daml/script/Runner.scala

```diff
@@ -0,0 +1,341 @@
+import com.digitalasset.ledger.api.v1.transaction.TreeEvent
+import com.digitalasset.ledger.api.v1.transaction_filter.{
+  Filters,
+  TransactionFilter,
+  InclusiveFilters
+}
+import com.digitalasset.ledger.api.v1.value.{Identifier => ApiIdentifier}
+import com.digitalasset.ledger.api.validation.ValueValidator
+import com.digitalasset.ledger.client.LedgerClient
+import com.digitalasset.platform.participant.util.LfEngineToApi.{
+  toApiIdentifier,
+  lfValueToApiRecord,
+  lfValueToApiValue
+}
+
+class Runner(dar: Dar[(PackageId, Package)], applicationId: ApplicationId) extends StrictLogging {
+
+  val darMap: Map[PackageId, Package] = dar.all.toMap
+  val compiler = Compiler(darMap)
+  val scriptModuleName = DottedName.assertFromString("Daml.Script")
+  val scriptPackageId: PackageId = dar.all
+    .find {
+      case (pkgId, pkg) => pkg.modules.contains(scriptModuleName)
+    }
+    .get
+    ._1
+  val stdlibPackageId =
+    dar.all
+      .find {
+        case (pkgId, pkg) =>
+          pkg.modules.contains(DottedName.assertFromString("DA.Internal.LF"))
+      }
+      .get
+      ._1
+
+  // We overwrite the definition of toLedgerValue with an identity function.
+  // This is a type error but Speedy doesn’t care about the types and the only thing we do
+  // with the result is convert it to ledger values/record so this is safe.
+  val definitionMap =
+    compiler.compilePackages(darMap.keys) +
+      (LfDefRef(
+        Identifier(
+          scriptPackageId,
+          QualifiedName(scriptModuleName, DottedName.assertFromString("fromLedgerValue")))) ->
+        SEMakeClo(Array(), 1, SEVar(1)))
+  val compiledPackages = PureCompiledPackages(darMap, definitionMap).right.get
+
+  def toLedgerRecord(v: SValue) = {
+    lfValueToApiRecord(
+      true,
+      v.toValue.mapContractId {
+        case rcoid: RelativeContractId =>
+          throw new RuntimeException(s"Unexpected contract id $rcoid")
+        case acoid: AbsoluteContractId => acoid
+      }
+    )
+  }
+
+  def toLedgerValue(v: SValue) = {
+    lfValueToApiValue(
+      true,
+      v.toValue.mapContractId {
+        case rcoid: RelativeContractId =>
+          throw new RuntimeException(s"Unexpected contract id $rcoid")
+        case acoid: AbsoluteContractId => acoid
+      }
+    )
+  }
+
+  def getApFields(fun: SValue): (SVariant, SVariant) = {
+    val extractTuple = SEMakeClo(
+      Array(),
+      2,
+      SEApp(
+        SEBuiltin(SBTupleCon(Name.Array(Name.assertFromString("a"), Name.assertFromString("b")))),
+        Array(SEVar(2), SEVar(1))))
+    val machine =
+      Speedy.Machine.fromSExpr(SEApp(SEValue(fun), Array(extractTuple)), false, compiledPackages)
+    while (!machine.isFinal) {
+      machine.step() match {
+        case SResultContinue => ()
+        case res => {
+          throw new RuntimeException(s"Unexpected speedy result $res")
+        }
+      }
+    }
+    val tuple = machine.toSValue.asInstanceOf[STuple]
```

Comment:
you should use match here, something like this:
```
machine.toSValue match {
  case (STuple(_, (a0: SVariant) :: (a1: SVariant) :: _) => (a0, a1)
  case => sys.error("something unexpected")
}
```
or  if you know that it is a tuple 2
```
machine.toSValue match {
  case (STuple(_, (a0: SVariant) :: (a1: SVariant) :: Nil) => (a0, a1)
  case => sys.error("something unexpected")
}
```
in any case using `match` is more idiomatic Scala, you should never use `asInstanceOf`.

Reply:
oops it is `java.util.ArrayList` you cannot match it like this, so you might want to convert it to Scala first. But in any case using `asInstanceOf` is discouraged unless you are hardcore Java programmer :)

Reply:
Yeah I’ll kill all the `asInstanceOf` stuff in a follow-up PR. That was just for prototyping.